File: python/python/glide/async_commands/server_modules/ft_options/ft_create_options.py
# Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
from abc import ABC, abstractmethod
from enum import Enum
from typing import List, Optional

from glide.async_commands.server_modules.ft_constants import FtCreateKeywords
from glide.constants import TEncodable


class FieldType(Enum):
    """
    All possible values for the data type of field identifier for the SCHEMA option.
    """

    TEXT = "TEXT"
    """
    If the field contains any blob of data.
    """
    TAG = "TAG"
    """
    If the field contains a tag field.     
    """
    NUMERIC = "NUMERIC"
    """
    If the field contains a number.
    """
    VECTOR = "VECTOR"
    """
    If the field is a vector field that supports vector search.
    """


class VectorAlgorithm(Enum):
    """
    Algorithm for vector type fields used for vector similarity search.
    """

    HNSW = "HNSW"
    """
    Hierarchical Navigable Small World algorithm.
    """
    FLAT = "FLAT"
    """
    Flat algorithm or the brute force algorithm.
    """


class DistanceMetricType(Enum):
    """
    The metric options for the distance in vector type field.
    """

    L2 = "L2"
    """
    Euclidean distance
    """
    IP = "IP"
    """
    Inner product
    """
    COSINE = "COSINE"
    """
    Cosine distance
    """


class VectorType(Enum):
    """
    Type type for the vector field type.
    """

    FLOAT32 = "FLOAT32"
    """
    FLOAT32 type of vector. The only supported type.
    """


class Field(ABC):
    """
    Abstract base class for defining fields in a schema.
    """

    @abstractmethod
    def __init__(
        self,
        name: TEncodable,
        type: FieldType,
        alias: Optional[str] = None,
    ):
        """
        Initialize a new field instance.

        Args:
            name (TEncodable): The name of the field.
            type (FieldType): The type of the field.
            alias (Optional[str]): An alias for the field.
        """
        self.name = name
        self.type = type
        self.alias = alias

    @abstractmethod
    def toArgs(self) -> List[TEncodable]:
        """
        Get the arguments representing the field.

        Returns:
            List[TEncodable]: A list of field arguments.
        """
        args = [self.name]
        if self.alias:
            args.extend([FtCreateKeywords.AS, self.alias])
        args.append(self.type.value)
        return args


class TextField(Field):
    """
    Class for defining text fields in a schema.
    """

    def __init__(self, name: TEncodable, alias: Optional[str] = None):
        """
        Initialize a new TextField instance.

        Args:
            name (TEncodable): The name of the text field.
            alias (Optional[str]): An alias for the field.
        """
        super().__init__(name, FieldType.TEXT, alias)

    def toArgs(self) -> List[TEncodable]:
        """
        Get the arguments representing the text field.

        Returns:
            List[TEncodable]: A list of text field arguments.
        """
        args = super().toArgs()
        return args


class TagField(Field):
    """
    Class for defining tag fields in a schema.
    """

    def __init__(
        self,
        name: TEncodable,
        alias: Optional[str] = None,
        separator: Optional[str] = None,
        case_sensitive: bool = False,
    ):
        """
        Initialize a new TagField instance.

        Args:
            name (TEncodable): The name of the tag field.
            alias (Optional[str]): An alias for the field.
            separator (Optional[str]): Specify how text in the attribute is split into individual tags. Must be a single character.
            case_sensitive (bool): Preserve the original letter cases of tags. If set to False, characters are converted to lowercase by default.
        """
        super().__init__(name, FieldType.TAG, alias)
        self.separator = separator
        self.case_sensitive = case_sensitive

    def toArgs(self) -> List[TEncodable]:
        """
        Get the arguments representing the tag field.

        Returns:
            List[TEncodable]: A list of tag field arguments.
        """
        args = super().toArgs()
        if self.separator:
            args.extend([FtCreateKeywords.SEPARATOR, self.separator])
        if self.case_sensitive:
            args.append(FtCreateKeywords.CASESENSITIVE)
        return args


class NumericField(Field):
    """
    Class for defining the numeric fields in a schema.
    """

    def __init__(self, name: TEncodable, alias: Optional[str] = None):
        """
        Initialize a new NumericField instance.

        Args:
            name (TEncodable): The name of the numeric field.
            alias (Optional[str]): An alias for the field.
        """
        super().__init__(name, FieldType.NUMERIC, alias)

    def toArgs(self) -> List[TEncodable]:
        """
        Get the arguments representing the numeric field.

        Returns:
            List[TEncodable]: A list of numeric field arguments.
        """
        args = super().toArgs()
        return args


class VectorFieldAttributes(ABC):
    """
    Abstract base class for defining vector field attributes to be used after the vector algorithm name.
    """

    @abstractmethod
    def __init__(self, dim: int, distance_metric: DistanceMetricType, type: VectorType):
        """
        Initialize a new vector field attributes instance.

        Args:
            dim (int): Number of dimensions in the vector.
            distance_metric (DistanceMetricType): The distance metric used in vector type field. Can be one of [L2 | IP | COSINE].
            type (VectorType): Vector type. The only supported type is FLOAT32.
        """
        self.dim = dim
        self.distance_metric = distance_metric
        self.type = type

    @abstractmethod
    def toArgs(self) -> List[str]:
        """
        Get the arguments to be used for the algorithm of the vector field.

        Returns:
            List[str]: A list of arguments.
        """
        args = []
        if self.dim:
            args.extend([FtCreateKeywords.DIM, str(self.dim)])
        if self.distance_metric:
            args.extend([FtCreateKeywords.DISTANCE_METRIC, self.distance_metric.name])
        if self.type:
            args.extend([FtCreateKeywords.TYPE, self.type.name])
        return args


class VectorFieldAttributesFlat(VectorFieldAttributes):
    """
    Get the arguments to be used for the FLAT algorithm of the vector field.
    """

    def __init__(
        self,
        dim: int,
        distance_metric: DistanceMetricType,
        type: VectorType,
        initial_cap: Optional[int] = None,
    ):
        """
        Initialize a new flat vector field attributes instance.

        Args:
            dim (int): Number of dimensions in the vector.
            distance_metric (DistanceMetricType): The distance metric used in vector type field. Can be one of [L2 | IP | COSINE].
            type (VectorType): Vector type. The only supported type is FLOAT32.
            initial_cap (Optional[int]): Initial vector capacity in the index affecting memory allocation size of the index. Defaults to 1024.
        """
        super().__init__(dim, distance_metric, type)
        self.initial_cap = initial_cap

    def toArgs(self) -> List[str]:
        """
        Get the arguments representing the vector field created with FLAT algorithm.

        Returns:
            List[str]: A list of FLAT algorithm type vector arguments.
        """
        args = super().toArgs()
        if self.initial_cap:
            args.extend([FtCreateKeywords.INITIAL_CAP, str(self.initial_cap)])
        return args


class VectorFieldAttributesHnsw(VectorFieldAttributes):
    """
    Get the arguments to be used for the HNSW algorithm of the vector field.
    """

    def __init__(
        self,
        dim: int,
        distance_metric: DistanceMetricType,
        type: VectorType,
        initial_cap: Optional[int] = None,
        m: Optional[int] = None,
        ef_contruction: Optional[int] = None,
        ef_runtime: Optional[int] = None,
    ):
        """
        Initialize a new TagField instance.

        Args:
            dim (int): Number of dimensions in the vector.
            distance_metric (DistanceMetricType): The distance metric used in vector type field. Can be one of [L2 | IP | COSINE].
            type (VectorType): Vector type. The only supported type is FLOAT32.
            initial_cap (Optional[int]): Initial vector capacity in the index affecting memory allocation size of the index. Defaults to 1024.
            m (Optional[int]): Number of maximum allowed outgoing edges for each node in the graph in each layer. Default is 16, maximum is 512.
            ef_contruction (Optional[int]): Controls the number of vectors examined during index construction. Default value is 200, Maximum value is 4096.
            ef_runtime (Optional[int]): Controls the number of vectors examined during query operations. Default value is 10, Maximum value is 4096.
        """
        super().__init__(dim, distance_metric, type)
        self.initial_cap = initial_cap
        self.m = m
        self.ef_contruction = ef_contruction
        self.ef_runtime = ef_runtime

    def toArgs(self) -> List[str]:
        """
        Get the arguments representing the vector field created with HSNW algorithm.

        Returns:
            List[str]: A list of HNSW algorithm type vector arguments.
        """
        args = super().toArgs()
        if self.initial_cap:
            args.extend([FtCreateKeywords.INITIAL_CAP, str(self.initial_cap)])
        if self.m:
            args.extend([FtCreateKeywords.M, str(self.m)])
        if self.ef_contruction:
            args.extend([FtCreateKeywords.EF_CONSTRUCTION, str(self.ef_contruction)])
        if self.ef_runtime:
            args.extend([FtCreateKeywords.EF_RUNTIME, str(self.ef_runtime)])
        return args


class VectorField(Field):
    """
    Class for defining vector field in a schema.
    """

    def __init__(
        self,
        name: TEncodable,
        algorithm: VectorAlgorithm,
        attributes: VectorFieldAttributes,
        alias: Optional[str] = None,
    ):
        """
        Initialize a new VectorField instance.

        Args:
            name (TEncodable): The name of the vector field.
            algorithm (VectorAlgorithm): The vector indexing algorithm.
            alias (Optional[str]): An alias for the field.
            attributes (VectorFieldAttributes): Additional attributes to be passed with the vector field after the algorithm name.
        """
        super().__init__(name, FieldType.VECTOR, alias)
        self.algorithm = algorithm
        self.attributes = attributes

    def toArgs(self) -> List[TEncodable]:
        """
        Get the arguments representing the vector field.

        Returns:
            List[TEncodable]: A list of vector field arguments.
        """
        args = super().toArgs()
        args.append(self.algorithm.value)
        if self.attributes:
            attribute_list = self.attributes.toArgs()
            args.append(str(len(attribute_list)))
            args.extend(attribute_list)
        return args


class DataType(Enum):
    """
    Options for the type of data for which the index is being created.
    """

    HASH = 1
    """
    If the created index will index HASH data.
    """
    JSON = 2
    """
    If the created index will index JSON document data.
    """


class FtCreateOptions:
    """
    This class represents the input options to be used in the [FT.CREATE] command.
    All fields in this class are optional inputs for [FT.CREATE].
    """

    def __init__(
        self,
        data_type: Optional[DataType] = None,
        prefixes: Optional[List[str]] = None,
    ):
        """
        Initialize the [FT.CREATE] optional fields.

        Args:
            data_type (Optional[DataType]): The type of data to be indexed using [FT.CREATE].
            prefixes (Optional[List[str]]): The prefix of the key to be indexed.
        """
        self.data_type = data_type
        self.prefixes = prefixes

    def toArgs(self) -> List[str]:
        """
        Get the optional arguments for the [FT.CREATE] command.

        Returns:
            List[str]:
                List of [FT.CREATE] optional agruments.
        """
        args = []
        if self.data_type:
            args.append(FtCreateKeywords.ON)
            args.append(self.data_type.name)
        if self.prefixes:
            args.append(FtCreateKeywords.PREFIX)
            args.append(str(len(self.prefixes)))
            for prefix in self.prefixes:
                args.append(prefix)
        return args

File: go/options/ft_create_options.go
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

package options

import (
	"errors"
	"strconv"

	"github.com/valkey-io/valkey-glide/go/v2/constants"
)

// fieldType is the internal wire type for a field in the index schema.
type fieldType string

const (
	fieldTypeText    fieldType = "TEXT"
	fieldTypeTag     fieldType = "TAG"
	fieldTypeNumeric fieldType = "NUMERIC"
	fieldTypeVector  fieldType = "VECTOR"
)

// vectorAlgorithm is the internal wire type for the vector indexing algorithm.
type vectorAlgorithm string

const (
	vectorAlgorithmHNSW vectorAlgorithm = "HNSW"
	vectorAlgorithmFlat vectorAlgorithm = "FLAT"
)

// Field is the interface implemented by all index schema field types.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type Field interface {
	// ToArgs returns the command arguments for this field.
	ToArgs() ([]string, error)
}

// baseField holds common field properties shared by all field types.
type baseField struct {
	Name     string
	Alias    string
	Sortable bool
}

func (f *baseField) baseArgs(ft fieldType) []string {
	args := []string{f.Name}
	if f.Alias != "" {
		args = append(args, "AS", f.Alias)
	}
	args = append(args, string(ft))
	return args
}

// TextField represents a full-text search field.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type TextField struct {
	baseField
	// NoStem disables stemming when indexing the field.
	NoStem bool
	// Weight declares the importance of this field when calculating result accuracy. Default is 1.
	Weight *float64
	// WithSuffixTrie keeps a suffix trie for the field to optimize contains and suffix queries.
	// Mutually exclusive with NoSuffixTrie.
	WithSuffixTrie bool
	// NoSuffixTrie disables the suffix trie for the field.
	// Mutually exclusive with WithSuffixTrie.
	NoSuffixTrie bool
}

// NewTextField creates a new TextField with the given name.
func NewTextField(name string) *TextField {
	return &TextField{baseField: baseField{Name: name}}
}

// SetAlias sets an alias for the field.
func (f *TextField) SetAlias(alias string) *TextField { f.Alias = alias; return f }

// SetSortable enables sorting on this field.
func (f *TextField) SetSortable(sortable bool) *TextField { f.Sortable = sortable; return f }

// SetNoStem disables stemming when indexing the field.
func (f *TextField) SetNoStem(nostem bool) *TextField { f.NoStem = nostem; return f }

// SetWeight sets the importance of this field when calculating result accuracy.
func (f *TextField) SetWeight(weight float64) *TextField { f.Weight = &weight; return f }

// SetWithSuffixTrie enables suffix trie for the field.
func (f *TextField) SetWithSuffixTrie(v bool) *TextField { f.WithSuffixTrie = v; return f }

// SetNoSuffixTrie disables suffix trie for the field.
func (f *TextField) SetNoSuffixTrie(v bool) *TextField { f.NoSuffixTrie = v; return f }

// ToArgs returns the command arguments for this TextField.
func (f *TextField) ToArgs() ([]string, error) {
	if f.WithSuffixTrie && f.NoSuffixTrie {
		return nil, errors.New("WithSuffixTrie and NoSuffixTrie are mutually exclusive")
	}
	args := f.baseArgs(fieldTypeText)
	if f.NoStem {
		args = append(args, "NOSTEM")
	}
	if f.Weight != nil {
		args = append(args, "WEIGHT", strconv.FormatFloat(*f.Weight, 'g', -1, 64))
	}
	if f.WithSuffixTrie {
		args = append(args, "WITHSUFFIXTRIE")
	} else if f.NoSuffixTrie {
		args = append(args, "NOSUFFIXTRIE")
	}
	if f.Sortable {
		args = append(args, "SORTABLE")
	}
	return args, nil
}

// TagField represents a tag field (a list of tags delimited by a separator character).
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type TagField struct {
	baseField
	// Separator specifies how text in the attribute is split into individual tags. Must be a single character.
	Separator string
	// CaseSensitive preserves the original letter cases of tags.
	CaseSensitive bool
}

// NewTagField creates a new TagField with the given name.
func NewTagField(name string) *TagField {
	return &TagField{baseField: baseField{Name: name}}
}

// SetAlias sets an alias for the field.
func (f *TagField) SetAlias(alias string) *TagField { f.Alias = alias; return f }

// SetSortable enables sorting on this field.
func (f *TagField) SetSortable(sortable bool) *TagField { f.Sortable = sortable; return f }

// SetSeparator sets the tag separator character.
func (f *TagField) SetSeparator(sep string) *TagField { f.Separator = sep; return f }

// SetCaseSensitive preserves original letter cases of tags.
func (f *TagField) SetCaseSensitive(v bool) *TagField { f.CaseSensitive = v; return f }

// ToArgs returns the command arguments for this TagField.
func (f *TagField) ToArgs() ([]string, error) {
	args := f.baseArgs(fieldTypeTag)
	if f.Separator != "" {
		args = append(args, "SEPARATOR", f.Separator)
	}
	if f.CaseSensitive {
		args = append(args, "CASESENSITIVE")
	}
	if f.Sortable {
		args = append(args, "SORTABLE")
	}
	return args, nil
}

// NumericField represents a numeric field.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type NumericField struct {
	baseField
}

// NewNumericField creates a new NumericField with the given name.
func NewNumericField(name string) *NumericField {
	return &NumericField{baseField: baseField{Name: name}}
}

// SetAlias sets an alias for the field.
func (f *NumericField) SetAlias(alias string) *NumericField { f.Alias = alias; return f }

// SetSortable enables sorting on this field.
func (f *NumericField) SetSortable(sortable bool) *NumericField { f.Sortable = sortable; return f }

// ToArgs returns the command arguments for this NumericField.
func (f *NumericField) ToArgs() ([]string, error) {
	args := f.baseArgs(fieldTypeNumeric)
	if f.Sortable {
		args = append(args, "SORTABLE")
	}
	return args, nil
}

// VectorFieldFlat represents a vector field using the FLAT (brute force) algorithm.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type VectorFieldFlat struct {
	baseField
	// DistanceMetric is the distance metric used in vector similarity search.
	DistanceMetric constants.DistanceMetric
	// Dimensions is the number of dimensions in the vector.
	Dimensions int
	// Type is the vector data type. Defaults to FLOAT32.
	Type constants.VectorDataType
	// InitialCap is the initial vector capacity in the index affecting memory allocation size.
	InitialCap *int
}

// NewVectorFieldFlat creates a new VectorFieldFlat with required parameters.
func NewVectorFieldFlat(name string, distanceMetric constants.DistanceMetric, dimensions int) *VectorFieldFlat {
	return &VectorFieldFlat{
		baseField:      baseField{Name: name},
		DistanceMetric: distanceMetric,
		Dimensions:     dimensions,
		Type:           constants.VectorDataTypeFloat32,
	}
}

// SetAlias sets an alias for the field.
func (f *VectorFieldFlat) SetAlias(alias string) *VectorFieldFlat { f.Alias = alias; return f }

// SetType sets the vector data type.
func (f *VectorFieldFlat) SetType(t constants.VectorDataType) *VectorFieldFlat {
	f.Type = t
	return f
}

// SetInitialCap sets the initial vector capacity in the index.
func (f *VectorFieldFlat) SetInitialCap(cap int) *VectorFieldFlat { f.InitialCap = &cap; return f }

// ToArgs returns the command arguments for this VectorFieldFlat.
func (f *VectorFieldFlat) ToArgs() ([]string, error) {
	args := f.baseArgs(fieldTypeVector)
	args = append(args, string(vectorAlgorithmFlat))
	attrs := f.buildAttrs()
	args = append(args, strconv.Itoa(len(attrs)))
	args = append(args, attrs...)
	return args, nil
}

func (f *VectorFieldFlat) buildAttrs() []string {
	attrs := []string{
		"DIM", strconv.Itoa(f.Dimensions),
		"DISTANCE_METRIC", string(f.DistanceMetric),
		"TYPE", string(f.Type),
	}
	if f.InitialCap != nil {
		attrs = append(attrs, "INITIAL_CAP", strconv.Itoa(*f.InitialCap))
	}
	return attrs
}

// VectorFieldHNSW represents a vector field using the HNSW algorithm.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type VectorFieldHNSW struct {
	baseField
	// DistanceMetric is the distance metric used in vector similarity search.
	DistanceMetric constants.DistanceMetric
	// Dimensions is the number of dimensions in the vector.
	Dimensions int
	// Type is the vector data type. Defaults to FLOAT32.
	Type constants.VectorDataType
	// InitialCap is the initial vector capacity in the index affecting memory allocation size.
	InitialCap *int
	// NumberOfEdges is the max number of outgoing edges for each node in the graph per layer.
	// Default is 16, maximum is 512. Equivalent to M in the module API.
	NumberOfEdges *int
	// VectorsExaminedOnConstruction controls the number of vectors examined during index construction.
	// Default is 200, maximum is 4096. Equivalent to EF_CONSTRUCTION in the module API.
	VectorsExaminedOnConstruction *int
	// VectorsExaminedOnRuntime controls the number of vectors examined during query operations.
	// Default is 10, maximum is 4096. Equivalent to EF_RUNTIME in the module API.
	VectorsExaminedOnRuntime *int
}

// NewVectorFieldHNSW creates a new VectorFieldHNSW with required parameters.
func NewVectorFieldHNSW(name string, distanceMetric constants.DistanceMetric, dimensions int) *VectorFieldHNSW {
	return &VectorFieldHNSW{
		baseField:      baseField{Name: name},
		DistanceMetric: distanceMetric,
		Dimensions:     dimensions,
		Type:           constants.VectorDataTypeFloat32,
	}
}

// SetAlias sets an alias for the field.
func (f *VectorFieldHNSW) SetAlias(alias string) *VectorFieldHNSW { f.Alias = alias; return f }

// SetType sets the vector data type.
func (f *VectorFieldHNSW) SetType(t constants.VectorDataType) *VectorFieldHNSW {
	f.Type = t
	return f
}

// SetInitialCap sets the initial vector capacity in the index.
func (f *VectorFieldHNSW) SetInitialCap(cap int) *VectorFieldHNSW { f.InitialCap = &cap; return f }

// SetNumberOfEdges sets the max number of outgoing edges per node (M parameter).
func (f *VectorFieldHNSW) SetNumberOfEdges(m int) *VectorFieldHNSW {
	f.NumberOfEdges = &m
	return f
}

// SetVectorsExaminedOnConstruction sets the EF_CONSTRUCTION parameter.
func (f *VectorFieldHNSW) SetVectorsExaminedOnConstruction(ef int) *VectorFieldHNSW {
	f.VectorsExaminedOnConstruction = &ef
	return f
}

// SetVectorsExaminedOnRuntime sets the EF_RUNTIME parameter.
func (f *VectorFieldHNSW) SetVectorsExaminedOnRuntime(ef int) *VectorFieldHNSW {
	f.VectorsExaminedOnRuntime = &ef
	return f
}

// ToArgs returns the command arguments for this VectorFieldHNSW.
func (f *VectorFieldHNSW) ToArgs() ([]string, error) {
	args := f.baseArgs(fieldTypeVector)
	args = append(args, string(vectorAlgorithmHNSW))
	attrs := f.buildAttrs()
	args = append(args, strconv.Itoa(len(attrs)))
	args = append(args, attrs...)
	return args, nil
}

func (f *VectorFieldHNSW) buildAttrs() []string {
	attrs := []string{
		"DIM", strconv.Itoa(f.Dimensions),
		"DISTANCE_METRIC", string(f.DistanceMetric),
		"TYPE", string(f.Type),
	}
	if f.InitialCap != nil {
		attrs = append(attrs, "INITIAL_CAP", strconv.Itoa(*f.InitialCap))
	}
	if f.NumberOfEdges != nil {
		attrs = append(attrs, "M", strconv.Itoa(*f.NumberOfEdges))
	}
	if f.VectorsExaminedOnConstruction != nil {
		attrs = append(attrs, "EF_CONSTRUCTION", strconv.Itoa(*f.VectorsExaminedOnConstruction))
	}
	if f.VectorsExaminedOnRuntime != nil {
		attrs = append(attrs, "EF_RUNTIME", strconv.Itoa(*f.VectorsExaminedOnRuntime))
	}
	return attrs
}

// FtCreateOptions holds optional arguments for the FT.CREATE command.
//
// See [valkey.io] for details.
//
// [valkey.io]: https://valkey.io/commands/ft.create/
type FtCreateOptions struct {
	// DataType is the index data type. If not set, a HASH index is created.
	DataType constants.IndexDataType
	// Prefixes is a list of key prefixes to index.
	Prefixes []string
	// Score is the default score for documents in the index. Default is 1.0.
	Score *float64
	// Language is the default language for documents in the index.
	Language string
	// SkipInitialScan skips scanning and indexing existing documents on index creation.
	SkipInitialScan bool
	// MinStemSize is the minimum word length to stem. Words shorter than this are not stemmed.
	MinStemSize *int
	// WithOffsets stores term offsets for document fields. Mutually exclusive with NoOffsets.
	WithOffsets bool
	// NoOffsets does not store term offsets. Mutually exclusive with WithOffsets.
	NoOffsets bool
	// NoStopWords disables stop-word filtering. Mutually exclusive with StopWords.
	NoStopWords bool
	// StopWords is a custom list of stop words. Mutually exclusive with NoStopWords.
	StopWords []string
	// Punctuation is a custom set of punctuation characters to use during tokenization.
	Punctuation string
}

// ToArgs returns the command arguments for FtCreateOptions.
func (o *FtCreateOptions) ToArgs() ([]string, error) {
	if o.WithOffsets && o.NoOffsets {
		return nil, errors.New("WithOffsets and NoOffsets are mutually exclusive")
	}
	if o.NoStopWords && len(o.StopWords) > 0 {
		return nil, errors.New("NoStopWords and StopWords are mutually exclusive")
	}
	args := []string{}
	if o.DataType != "" {
		args = append(args, "ON", string(o.DataType))
	}
	if len(o.Prefixes) > 0 {
		args = append(args, "PREFIX", strconv.Itoa(len(o.Prefixes)))
		args = append(args, o.Prefixes...)
	}
	if o.Score != nil {
		args = append(args, "SCORE", strconv.FormatFloat(*o.Score, 'g', -1, 64))
	}
	if o.Language != "" {
		args = append(args, "LANGUAGE", o.Language)
	}
	if o.SkipInitialScan {
		args = append(args, "SKIPINITIALSCAN")
	}
	if o.MinStemSize != nil {
		args = append(args, "MINSTEMSIZE", strconv.Itoa(*o.MinStemSize))
	}
	if o.WithOffsets {
		args = append(args, "WITHOFFSETS")
	} else if o.NoOffsets {
		args = append(args, "NOOFFSETS")
	}
	if o.NoStopWords {
		args = append(args, "NOSTOPWORDS")
	} else if len(o.StopWords) > 0 {
		args = append(args, "STOPWORDS", strconv.Itoa(len(o.StopWords)))
		args = append(args, o.StopWords...)
	}
	if o.Punctuation != "" {
		args = append(args, "PUNCTUATION", o.Punctuation)
	}
	return args, nil
}
